src/kudu/consensus/routing.cc

```diff
@@ -507,6 +507,139 @@ string DurableRoutingTable::LogPrefix() const {
+  // directly to all such peers
+  // 2. A peer which is in the same region as the node that is shipping messages
+  // will not be proxied i.e all peers in the same region as the 'source' will
+  // get messages directly from the 'source'
+  // 3. A peer which is in a region without and valid 'proxy peer' will recieve
+  // messages directly from the 'source'
+  std::unordered_map<std::string, std::string> dst_to_proxy_map;
+  for (const RaftPeerPB& peer : raft_config.peers()) {
+    std::string peer_region = peer.attrs().region();
+    if (peer.attrs().backing_db_present()) {
+      // Peers that have a backing database are not proxied (rule #1)
+      continue;
+    } else {
+      const auto& proxy_peer_uuid = region_proxy_peer_map.find(peer_region);
+      if (proxy_peer_uuid == region_proxy_peer_map.end() ||
+          peer_region == local_peer_region) {
+        // Region without a valid 'proxy' peer or peers that are in the same
+        // region as this peer are not proxied (rule #2 and #3)
+        continue;
+      } else {
+        // Add a new edge into the topology
+        ProxyEdgePB* proxy_edge = proxy_topology.add_proxy_edges();
+        proxy_edge->set_peer_uuid(peer.permanent_uuid());
+
+        // 'peer' will be proxied from 'proxy_peer_uuid'
+        proxy_edge->set_proxy_from_uuid(proxy_peer_uuid->second);
+        dst_to_proxy_map.emplace(peer.permanent_uuid(), proxy_peer_uuid->second);
+      }
+    }
+  }
+
+  std::lock_guard<RWMutex> l(lock_);
+  proxy_topology_ = std::move(proxy_topology);
+  dst_to_proxy_map_ = std::move(dst_to_proxy_map);
+  peer_region_map_ = std::move(peer_region_map);
+  raft_config_ = std::move(raft_config);
+
+  for (const auto& proxy : dst_to_proxy_map_) {
+    LOG(INFO) << local_peer_region << " : " << proxy.second << " -> " << proxy.first;
+  }
+
+  return Status::OK();
+}
+
+Status SimpleRegionRoutingTable::NextHop(const std::string& src_uuid,
+                                         const std::string& dest_uuid,
+                                         std::string* next_hop) const {
+  shared_lock<RWMutex> l(lock_);
+  const auto& proxy_uuid = dst_to_proxy_map_.find(dest_uuid);
+  if (proxy_uuid == dst_to_proxy_map_.end()) {
+    // Could not find this destination, route directly to the destination
+    *next_hop = dest_uuid;
+    return Status::OK();
+  }
+
+  *next_hop = proxy_uuid->second;
+  return Status::OK();
+}
+
+Status SimpleRegionRoutingTable::UpdateProxyTopology(
+    ProxyTopologyPB proxy_topology) {
+  // SimpleRegionRoutingTable uses config to update proxy maps. Hence cannot
+  // update topology directly.
+  // TODO: provide a way to override designated per-region proxy peer
+  return Status::OK();
+}
+
+ProxyTopologyPB SimpleRegionRoutingTable::GetProxyTopology() const {
+  shared_lock<RWMutex> l(lock_);
+  return proxy_topology_;
+}
+
+Status SimpleRegionRoutingTable::UpdateRaftConfig(RaftConfigPB raft_config) {
+  return RebuildProxyTopology(std::move(raft_config));
```

Comment:
Where are the UpdateRaftConfig or RebuildProxyTopology functions being called?
Has this been integrated into Config change or is this in a later diff?

Is there a possibility that in an add operation the recently added member becomes the proxy
node?

Reply:
Discussed offline. Routing table's UpdateRaftConfig() is called whenever a new config is proposed/committed/aborted. Hence the topology gets updated there automatically.